### PR TITLE
Changes to signatures

### LINKED ISSUE #3: Adapt signature files to syntax described in issue #2
idea behind see end of issue #2

in short:

ipSignature is an array of ips.
requestSignature can have any of the following keys: method, status, uri. and body. At least one of those keys must be provided, but others are optional and get treated as AND conditions.

A possible signature file would then look like:

{
  "ipSignature" : [ "^10\\.0\\.0\\.10$", "^192\\.168\\.0\\.255$" ],

  "requestSignature" : [
       {
          "uri" : ".*\\.php"
        } ,
        {
           "method" : "GET",
           "body" : ".+"
        },
        {
           "status" : "[45]\\d\\d"
        }
   ]
}

### PR DESCRIPTION
**Pull request by geo-bert**
I deleted the old signature files and added a new one that matches the description in #3. I also added detailed documentation about creating a signature. I linked to the corresponding file in `Overview.md`.
As sources for the signatures I used some signatures we observed, ones I thought of myself, [this](https://www.researchgate.net/publication/320476110_Cybersecurity_and_Network_Forensics_Analysis_of_Malicious_Traffic_towards_a_Honeynet_with_Deep_Packet_Inspection) article and for example IPs [this](https://www.projecthoneypot.org/list_of_ips.php) website.

Additionally for testing purposes I used [this](https://reqbin.com/) website to generate HTTP requests.

This closes #3.

### CODE REVIEW

File: doc/Overview.md [hunk 1]
```diff
@@ -1,10 +1,12 @@
 # Overview
+
```
Comment (mhatzl) [inline doc/Overview.md:2]: Are spaces after a header convention on github?

File: signatures/example_signatures.json [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+            "uri": "^\/index.php$"
+        },
+        {
+            "method": "^GET$",
+            "uri": "^\/index.html$"
+        },
+        {
+            "method": "^GET$",
+            "uri": "^\/index.php$"
+        },
```
Comment (mhatzl) [inline signatures/example_signatures.json:21]: Duplicate to the first

File: signatures/example_signatures.json [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        "^176\\.191\\.129\\.\\d{1,3}$",
+        "^85\\.90\\.156\\.157$",
+        "^161\\.35\\.27\\.209$",
+        "^134\\.119\\.219\\.71$",
+        "^103\\.227\\.8\\.42$"
+    ],
+    "requestSignature": [
+        {
+            "method": "^GET$",
+            "uri": "^\/index.php$"
```
Comment (mhatzl) [inline signatures/example_signatures.json:12]: I think you need to write "\\/index.php" for the match to work, since one \ gets removed I think when converting to JSON.
Have you tried to match against one of those signatures?

File: signatures/example_signatures.json [hunk 4]
```diff
@@ -0,0 +1,90 @@
+{
+    "ipSignature": [
+        "^176\\.191\\.129\\.\\d{1,3}$",
```
Comment (mhatzl) [inline signatures/example_signatures.json:3]: Currently, inside web_ids the ip-field inside ipSignatures-objects is used. Like it was in the original signature file.
This might even be better than having an array (as mentioned in #2)
Keeping the ipSignature as object like before, gives us the option to provide a type field to define, if the ip match is of type INFO, WARNING, ...
Since we only parse fields known to us, one could for example add a msg-field to provide additional information for a certain match.

So in summary:
I would keep the ipSignature in the form

~~~
{
   "ip" : "<some ip to match>",
   "type" : "<INFO | WARNING | ...>,
   "msg" : "Optional message field that we ignore"
}
~~~

File: doc/Signatures.md [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+- New signatures are not loaded unless web_ids is rebooted!
+- Characters need to be escaped properly.
+- If you do not plan to add one kind of signature, leave the array empty
+
+## IP Signatures
+
+`ipSignature` is an array of Regex strings that represent certain IPs or IP ranges. For example, an entry might look like this:
+
+```json
+^123\\.456\\.789\\.012$
```
Comment (mhatzl) [inline doc/Signatures.md:27]: might be changed. See comment in signature file below

**Review by mhatzl — COMMENTED** [review]
Some very nice signature examples.
Your original ipSignature was already good, sorry for the confusion

File: signatures/example_signatures.json (hunk 2 above)
Comment (geo-bert) [reply signatures/example_signatures.json:21]: Will resolve

File: doc/Overview.md (hunk 1 above)
Comment (mhatzl) [reply doc/Overview.md:2]: ah ok :D what do you prefer?

File: signatures/example_signatures.json (hunk 4 above)
Comment (geo-bert) [reply signatures/example_signatures.json:3]: Ok. Will change this

File: doc/Overview.md (hunk 1 above)
Comment (geo-bert) [reply doc/Overview.md:2]: I can't easily turn of the autoformatter for markdown files so I'd like to keep this like it is now.

File: doc/Signatures.md (hunk 5 above)
Comment (geo-bert) [reply doc/Signatures.md:27]: Will be changed

File: signatures/example_signatures.json (hunk 4 above)
Comment (geo-bert) [reply signatures/example_signatures.json:3]: @mhatzl Is it ok to just include an IP key in the object in the example file and the documentation for now? Including the other keys we have not yet implemented would be bad programming practice.

File: signatures/example_signatures.json (hunk 3 above)
Comment (geo-bert) [reply signatures/example_signatures.json:12]: I don't exactly get what you mean. The following code yields `true` for me:
```js
content = JSON.parse(fs.readFileSync("../signatures/example_signatures.json"));
console.log("/index.php".match(content.requestSignature[0].uri) !== null);
```
And if I try to write `"\/index.php"` my autoformatter removes the backslash. Is this what you mean?

File: signatures/example_signatures.json (hunk 4 above)
Comment (mhatzl) [reply signatures/example_signatures.json:3]: Yes thats good

File: signatures/example_signatures.json (hunk 3 above)
Comment (mhatzl) [reply signatures/example_signatures.json:22]: ok I think one backslash got removed in my comment. In the test yesterday, I had to use two backslashes before /index.php.
Otherwise web_ids wouldn't match against a log containing index.php.
But to generally avoid this slash/backslash escaping, we could just write "index.php", since we don't care at what position this string is in the request

Comment (mhatzl) [reply signatures/example_signatures.json:22]: Ok, I have tried it now on the server, and it seems to work now with all variations '\\/index.php', '\/index.php' and '/index.php'
Looking at the regex documentation [here](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions), forward slashes are used to define start and end of a regex statement. I guess internally the characters are slashes are escaped correctly when executing the match.

In this case I woul remove backslashes that escape forward slashes, as you already mentioned.

Comment (geo-bert) [reply signatures/example_signatures.json:22]: So do I need to remove them from the signature file aswell?

Comment (mhatzl) [reply signatures/example_signatures.json:22]: Not sure. The request is still matched, if we keep the signature as is.
It is just about the readability and to see, that you don't need to escape a forward slash.

For later users, it is probably better to have examples with no backslash escaping.

Comment (geo-bert) [reply signatures/example_signatures.json:22]: Ok, I will replace the escaped slashes with normal slashes

**Review by mhatzl — APPROVED** [review]
Looks good 👍
